src/app-layer-ftp.c

```diff
@@ -374,6 +378,55 @@ static void FtpTransferCmdFree(void *data)
     FTPFree(cmd, sizeof(struct FtpTransferCmd));
 }
 
+/**
+ * \brief This function is extract a port number from the command input line for IPv6 FTP usage
+ * \param input input line of the command
+ * \param input_len length of the request
+ *
+ * \retval 0 if a port number could not be extracted; otherwise, thed ynamic port number
+ */
+static uint16_t FTPGetV6PortNumber(uint8_t *input, uint32_t input_len)
+{
+    uint8_t *ptr = memrchr(input, '|', input_len);
+    if (ptr == NULL) {
+        return 0;
+    }
+
+    int n_length =  ptr - input - 1;
+    if (n_length < 4)
+        return 0;
+
+    ptr = memrchr(input, '|', n_length);
+    if (ptr == NULL)
+        return 0;
+
+    return atoi((char *)ptr + 1);
+}
+
+/**
+ * \brief This function is extract a port number from the command input line for IPv4 FTP usage
+ * \param input input line of the command
+ * \param input_len length of the request
+ *
+ * \retval 0 if a port number could not be extracted; otherwise, thed ynamic port number
+ */
+static int FTPGetV4PortNumber(uint8_t *input, uint32_t input_len)
```

Comment:
should we have validation in place on: numbers non-negative, values not leading to overflow, no missing values?

Reply:
The port number is  an unsigned 16-bite value -- all values other than 0 are valid. I'm not following the "no missing values" as a single value, not a range or other sequence, is being parsed.

Reply:
My point is that it makes sense to have validation that the values are valid. If the command badly formatted we should set an event and not create the expectation.